File: manager/shared/src/main/java/org/openremote/manager/shared/asset/Asset.java
/*
 * Copyright 2016, OpenRemote Inc.
 *
 * See the CONTRIBUTORS.txt file in the distribution for a
 * full listing of individual contributors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.openremote.manager.shared.asset;

import elemental.json.JsonObject;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Arrays;
import java.util.Date;

import static org.openremote.manager.shared.Constants.PERSISTENCE_JSON_OBJECT_TYPE;
import static org.openremote.manager.shared.Constants.PERSISTENCE_UNIQUE_ID_GENERATOR;

@MappedSuperclass
@Table(name = "ASSET")
public class Asset {

    @Id
    @Column(name = "ID", length = 22)
    @GeneratedValue(generator = PERSISTENCE_UNIQUE_ID_GENERATOR)
    protected String id;

    @Version
    @Column(name = "OBJ_VERSION", nullable = false)
    protected long version;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "CREATED_ON", updatable = false, nullable = false)
    @org.hibernate.annotations.CreationTimestamp
    protected Date createdOn = new Date();

    @NotNull
    @Column(name = "NAME", nullable = false)
    protected String name;

    @NotNull
    @Column(name = "ASSET_TYPE", nullable = false)
    protected String type;

    @Column(name = "PARENT_ID")
    protected String parentId;

    @Column(name = "ATTRIBUTES", columnDefinition = "jsonb")
    @org.hibernate.annotations.Type(type = PERSISTENCE_JSON_OBJECT_TYPE)
    protected JsonObject attributes;

    @Transient
    protected String[] path;

    @Transient
    protected double[] coordinates;

    public Asset() {
    }

    public Asset(String name, String type) {
        this.name = name;
        this.type = type;
    }

    public Asset(Asset parent) {
        this.parentId = parent.getId();
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public long getVersion() {
        return version;
    }

    public Date getCreatedOn() {
        return createdOn;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public void setType(AssetType type) {
        setType(type.getValue());
    }

    public String getParentId() {
        return parentId;
    }

    public void setParentId(String parentId) {
        this.parentId = parentId;
    }

    public JsonObject getAttributes() {
        return attributes;
    }

    public void setAttributes(JsonObject attributes) {
        this.attributes = attributes;
    }

    public String[] getPath() {
        return path;
    }

    public void setPath(String[] path) {
        this.path = path;
    }

    public double[] getCoordinates() {
        return coordinates;
    }

    public void setCoordinates(double... coordinates) {
        this.coordinates = coordinates;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "{" +
            "id='" + id + '\'' +
            ", name='" + name + '\'' +
            ", type ='" + type + '\'' +
            ", parent ='" + parentId + '\'' +
            ", path ='" + Arrays.toString(path) + '\'' +
            ", coordinates ='" + Arrays.toString(coordinates) + '\'' +
            ", attributes='" + (attributes != null ? attributes.toJson() : "null") + '\'' +
            '}';
    }

}

File: ui/component/or-asset-viewer/test/fixtures/data/asset.ts
/*
 * Copyright 2026, OpenRemote Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */
import type { Asset } from "@openremote/model";

export const notes = { name: "notes", type: "text" };
export const location = { name: "location", type: "GEO_JSONPoint" };
export const commonAttrs = { notes, location };
export const validAsset: Asset = {
  id: "validAsset",
  name: "Thing",
  realm: "master",
  type: "ThingAsset",
  attributes: { ...commonAttrs },
};
export const configuredAsset: Asset = {
    id: "configuredAsset",
    name: "Configured Thing",
    realm: "master",
    type: "ThingAsset",
    attributes: {
        notes: { ...notes, meta: { readOnly: true } },
        model: { name: "model", type: "text", meta: { label: "Model" } },
        location,
    },
};
export const invalidAsset: Asset = {
  id: "invalidAsset",
  name: "Thing",
  realm: "master",
  type: "ThingAsset",
  attributes: { ...commonAttrs, invalid: { name: "invalid", type: "integer" } },
};
